### DISCUSSION TITLE
How to do tessellation?

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3752] [Q&A] [answered]
Suppose I need to do some 2D tessellation (i.e., take a planar polygon that may not be convex and may have holes, and break it into triangles).  Up to now, I've done that using `gluNewTess` and friends.  If I'm getting rid of OpenGL and leaving all GPU access up to Filament, how would you suggest I do tessellation?  Assuming that this capability is not hidden in Filament, can anyone suggest open-source C++ software, hopefully with a fairly liberal license, that can do it on the CPU?

**Comment by romainguy** [comment]
gluNewTess is software based anyway. You should look for libraries that implement Delaunay triangulation. Maybe OpenSubdiv can do this as well?

**Comment by kpeeters** [comment]
`Earcut` does a pretty reasonable job, and is just a single header. See https://github.com/mapbox/earcut.hpp.

**Reply by jwwalker** [reply]
I tried Earcut, and it works for simple cases, but in complex cases it does not use the even-odd rule for inside vs. outside that I got with glu.

**Reply by prideout** [reply]
If you need even-odd support, maybe look at https://github.com/memononen/libtess2, written by the author of nanovg.

**Comment by jwwalker** [comment, marked as answer]
Thanks for the ideas, folks.  I think what I'll do is use the glu tessellation code from Mesa, which should maximize compatibility.
